File: Hackerrank_Strings.java
//Hackerrank_Strings.java

// // This exercise is to test your understanding of Java Strings. A sample String declaration:

// // String myString = "Hello World!"
// // The elements of a String are called characters. The number of characters in a String is called the length, and it can be retrieved with the String.length() method.

// // Given two strings of lowercase English letters,  A and B, perform the following operations:

// // Sum the lengths of A and B.
// // Determine if A is lexicographically larger than B (i.e.: does B  come before A in the dictionary?).
// // Capitalize the first letter in A and B and print them on a single line, separated by a space.



// import java.util.*;

// public class Solution {
//     public static void main(String[] args) {
//         Scanner sc = new Scanner(System.in);
//         String A = sc.next();
//         String B = sc.next();
//         sc.close();
        
//         // Calculate sum of lengths
//         int sumLengths = A.length() + B.length();
        
//         // Compare lexicographical order
//         String lexComparison = (A.compareTo(B) > 0) ? "Yes" : "No";
        
//         // Capitalize first letter of A and B
//         String capitalizedA = A.substring(0, 1).toUpperCase() + A.substring(1);
//         String capitalizedB = B.substring(0, 1).toUpperCase() + B.substring(1);
        
//         // Print results
//         System.out.println(sumLengths);
//         System.out.println(lexComparison);
//         System.out.println(capitalizedA + " " + capitalizedB);
//     }
// }





// // Given a string,s , and two indices, start and end, print a substring consisting of all characters in the inclusive range from  start to end-1. You'll find the String class' substring method helpful in completing this challenge.
// // Input Format
// // The first line contains a single string denoting s.
// // The second line contains two space-separated integers denoting the respective values of start and end.

// // Constraints
// // 1<=|s|<=100
// // 0<=start<end<=n

// // String s consists of English alphabetic letters (i.e.,[a-zA-Z] ) only.


// import java.util.*;

// public class Solution {
//     public static void main(String[] args) {
//         Scanner scanner = new Scanner(System.in);
        
//         // Read input string
//         String s = scanner.next();
        
//         // Read start and end indices
//         int start = scanner.nextInt();
//         int end = scanner.nextInt();
        
//         scanner.close();
        
//         // Extract substring from s
//         String substring = s.substring(start, end);
        
//         // Print the substring
//         System.out.println(substring);
//     }
// }




// // A substring of a string is a contiguous block of characters in the string. For example, the substrings of abc are a, b, c, ab, bc, and abc.
// // Given a string,s , and an integer, k, complete the function so that it finds the lexicographically smallest and largest substrings of length k.

// // Function Description

// // Complete the getSmallestAndLargest function in the editor below.

// // getSmallestAndLargest has the following parameters:

// // string s: a string
// // int k: the length of the substrings to find

// // Returns
// // string: the string ' + "\n" + ' where and are the two substrings

// // Input Format
// // The first line contains a string denoting s.
// // The second line contains an integer denoting k.


// import java.util.Scanner;

// public class Solution {
//     public static String getSmallestAndLargest(String s, int k) {
//         String smallest = s.substring(0, k);
//         String largest = s.substring(0, k);
        
//         for (int i = 1; i <= s.length() - k; i++) {
//             String currentSubstring = s.substring(i, i + k);
//             if (currentSubstring.compareTo(smallest) < 0) {
//                 smallest = currentSubstring;
//             }
//             if (currentSubstring.compareTo(largest) > 0) {
//                 largest = currentSubstring;
//             }
//         }
        
//         return smallest + "\n" + largest;
//     }


//     public static void main(String[] args) {
//         Scanner scan = new Scanner(System.in);
//         String s = scan.next();
//         int k = scan.nextInt();
//         scan.close();
      
//         System.out.println(getSmallestAndLargest(s, k));
//     }
// }






// // Using Regex, we can easily match or search for patterns in a text. Before searching for a pattern, we have to specify one using some well-defined syntax.
// // In this problem, you are given a pattern. You have to check whether the syntax of the given pattern is valid.
// // Note: In this problem, a regex is only valid if you can compile it using the Pattern.compile method.

// // Input Format
// // The first line of input contains an integer N, denoting the number of test cases. The next N lines contain a string of any printable characters representing the pattern of a regex.

// // Output Format
// // For each test case, print Valid if the syntax of the given pattern is correct. Otherwise, print Invalid. Do not print the quotes.

// // Sample Input
// // 3
// // ([A-Z])(.+)
// // [AZ[a-z](a-z)
// // batcatpat(nat

// // Sample Output
// // Valid
// // Invalid
// // Invalid




// import java.util.Scanner;
// import java.util.regex.*;

// public class Solution {
//     public static void main(String[] args) {
//         Scanner in = new Scanner(System.in);
//         int testCases = Integer.parseInt(in.nextLine());
//         while (testCases > 0) {
//             String pattern = in.nextLine();
//             try {
//                 Pattern.compile(pattern);
//                 System.out.println("Valid");
//             } catch (PatternSyntaxException e) {
//                 System.out.println("Invalid");
//             }
//             testCases--;
//         }
//         in.close();
//     }
// }





// // Write a class called MyRegex which will contain a string pattern. You need to write a regular expression and assign it to the pattern such that it can be used to validate an IP address. Use the following definition of an IP address:
// // IP address is a string in the form "A.B.C.D", where the value of A, B, C, and D may range from 0 to 255. Leading zeros are allowed. The length of A, B, C, or D can't be greater than 3.

// // Some valid IP address:
// // 000.12.12.034
// // 121.234.12.12
// // 23.45.12.56

// // Some invalid IP address:
// // 000.12.234.23.23
// // 666.666.23.23
// // .213.123.23.32
// // 23.45.22.32.
// // I.Am.not.an.ip
// // In this problem you will be provided strings containing any combination of ASCII characters. You have to write a regular expression to find the valid IPs.

// // Just write the MyRegex class which contains a String pattern. The string should contain the correct regular expression.
// // (MyRegex class MUST NOT be public)

// // Sample Input
// // 000.12.12.034
// // 121.234.12.12
// // 23.45.12.56
// // 00.12.123.123123.123
// // 122.23
// // Hello.IP

// // Sample Output
// // true
// // true
// // true
// // false
// // false
// // false




// import java.util.regex.Matcher;
// import java.util.regex.Pattern;
// import java.util.Scanner;

// class Solution{
//     public static void main(String[] args){
//         Scanner in = new Scanner(System.in);
//         while(in.hasNext()){
//             String IP = in.next();
//             System.out.println(IP.matches(new MyRegex().pattern));
//         }

//     }
// }


// class MyRegex {
//     String pattern;

//     public MyRegex() {
//         // Regular expression for a valid IP address
//         String zeroTo255 = "((25[0-5])|(2[0-4][0-9])|([01]?[0-9][0-9]?)|0)";
//         this.pattern = zeroTo255 + "\\." + zeroTo255 + "\\." + zeroTo255 + "\\." + zeroTo255;
//     }
// }

File: hackerrank_strings.py
# # You are given a string and your task is to swap cases. In other words,
# # convert all lowercase letters to uppercase letters and vice versa.

# # For Example:
# # Www.HackerRank.com → wWW.hACKERrANK.COM
# # Pythonist 2 → pYTHONIST 2  
# # Function Description

# # Complete the swap_case function in the editor below. swap_case has the following parameters:
# # string s: the string to modify
# # Returns
# # string: the modified string
# # Input Format
# # A single line containing a string s .


# def swap_case(s):
#     return s.swapcase()

# if __name__ == '__main__':
#     s = input()
#     result = swap_case(s)
#     print(result)



# # In Python, a string can be split on a delimiter.
# # Example:
# # >>> a = "this is a string"
# # >>> a = a.split(" ") # a is converted to a list of strings. 
# # >>> print a
# # ['this', 'is', 'a', 'string']
# # Joining a string is simple:

# # >>> a = "-".join(a)
# # >>> print a
# # this-is-a-string 



# # Task
# # You are given a string. Split the string on a " " (space) delimiter and join using a - hyphen.

# # Function Description
# # Complete the split_and_join function in the editor below.
# # split_and_join has the following parameters:
# # string line: a string of space-separated words
# # Returns

# # string: the resulting string
# # Input Format
# # The one line contains a string consisting of space separated words.

# # Sample Input
# # this is a string   
# # Sample Output
# # this-is-a-string


# def split_and_join(line):
#     words = line.split(" ")
#     result = "-".join(words)
#     return result

# if __name__ == '__main__':
#     line = input("Enter value: ")
#     result = split_and_join(line)
#     print(result)



# # You are given the firstname and lastname of a person on two different lines. Your task is to read them and print the following:
# # Hello firstname lastname! You just delved into python.

# # Function Description
# # Complete the print_full_name function in the editor below.
# # print_full_name has the following parameters:
# # string first: the first name
# # string last: the last name
# # Prints
# # string: 'Hello firstname lastname ! You just delved into python' where firstname and lastname are replaced with first and last.

# # Input Format
# # The first line contains the first name, and the second line contains the last name.

# # Constraints
# # The length of the first and last names are each ≤ 10


# def print_full_name(first, last):
#     print(f"Hello {first} {last}! You just delved into python.")
    
# if __name__ == '__main__':
#     first_name = input("Enter firstname: ")
#     last_name = input("Enter lastname:  ")
#     print_full_name(first_name, last_name)



# # We have seen that lists are mutable (they can be changed), and tuples are immutable (they cannot be changed).

# # Let's try to understand this with an example.
# # You are given an immutable string, and you want to make changes to it.

# # Example
# # >>> string = "abracadabra"
# # You can access an index by:

# # >>> print string[5]
# # a
# # What if you would like to assign a value?

# # >>> string[5] = 'k' 
# # Traceback (most recent call last):
# #   File "<stdin>", line 1, in <module>
# # TypeError: 'str' object does not support item assignment
# # How would you approach this?

# # One solution is to convert the string to a list and then change the value.
# # Example

# # >>> string = "abracadabra"
# # >>> l = list(string)
# # >>> l[5] = 'k'
# # >>> string = ''.join(l)
# # >>> print string
# # abrackdabra\


# def mutate_string(string, position, character):
#     string_list = list(string)
#     string_list[position] = character
#     return ''.join(string_list)

# if __name__ == '__main__':
#     s = input()
#     i, c = input().split()
#     s_new = mutate_string(s, int(i), c)
#     print(s_new)

#     OR

# def mutate_string(string: str, position: int, character: str) -> str:
#     before = string[:position]
#     new_char = character
#     after = string[position + 1:]
#     return before + new_char + after

# if __name__ == '__main__':
#     original_string = input("Enter the original string: ")
#     position = int(input("Enter the position to change (0-based index): "))
#     new_character = input("Enter the new character: ")
    
#     new_string = mutate_string(original_string, position, new_character)
#     print(new_string)



# # In this challenge, the user enters a string and a substring. You have to print the number of times that the substring occurs in the given string.
# #  String traversal will take place from left to right, not from right to left.
# # NOTE: String letters are case-sensitive.

# # Input Format
# # The first line of input contains the original string. The next line contains the substring.
# # Constraints
# # 1<=len(string) <= 200
# # Each character in the string is an ascii character.


# def count_substring(string, sub_string):
#     count = 0
#     sub_len = len(sub_string)
    
#     # Loop through the string
#     for i in range(len(string) - sub_len + 1):
#         # Check if the substring matches the part of the string
#         if string[i:i + sub_len] == sub_string:
#             count += 1
            
#     return count

# if __name__ == '__main__':
#     string = input().strip()
#     sub_string = input().strip()
    
#     count = count_substring(string, sub_string)
#     print(count)


# # Task

# # You are given a string .
# # Your task is to find out if the string  contains: alphanumeric characters, alphabetical characters, digits, lowercase and uppercase characters.

# # Input Format
# # A single line containing a string .

# # Constraints
# # 0<len(S) < 100

# # Output Format
# # In the first line, print True if  has any alphanumeric characters. Otherwise, print False.
# # In the second line, print True if  has any alphabetical characters. Otherwise, print False.
# # In the third line, print True if  has any digits. Otherwise, print False.
# # In the fourth line, print True if  has any lowercase characters. Otherwise, print False.
# # In the fifth line, print True if  has any uppercase characters. Otherwise, print False.

# # Sample Input

# # qA2
# # Sample Output

# # True
# # True
# # True
# # True
# # True




# if __name__ == '__main__':
#     s = input().strip()  # Input the string and strip any extra whitespace
    
#     # Initialize variables to store the results of checks
#     has_alnum = any(c.isalnum() for c in s)
#     has_alpha = any(c.isalpha() for c in s)
#     has_digit = any(c.isdigit() for c in s)
#     has_lower = any(c.islower() for c in s)
#     has_upper = any(c.isupper() for c in s)
    
#     # Print the results as required
#     print(has_alnum)
#     print(has_alpha)
#     print(has_digit)
#     print(has_lower)
#     print(has_upper)



# # You are given a string S and width w.
# # Your task is to wrap the string into a paragraph of width .

# # Function Description
# # Complete the wrap function in the editor below.
# # wrap has the following parameters:

# # string string: a long string
# # int max_width: the width to wrap to
# # Returns

# # string: a single string with newline characters ('\n') where the breaks should be



# import textwrap

# def wrap(string, max_width):
#     wrapped_lines = []
#     for i in range(0, len(string), max_width):
#         wrapped_lines.append(string[i:i + max_width])
#     return '\n'.join(wrapped_lines)


# if __name__ == '__main__':
#     string, max_width = input(), int(input())
#     result = wrap(string, max_width)
#     print(result)



# # Mr. Vincent works in a door mat manufacturing company. One day, he designed a new door mat with the following specifications:

# # Mat size must be N X M. (N is an odd natural number, and M is  3 times N.)
# # The design should have 'WELCOME' written in the center.
# # The design pattern should only use |, . and - characters.



# def print_door_mat(N, M):
#     # Top part
#     for i in range(N // 2):
#         pattern = '.|.' * (2 * i + 1)
#         print(pattern.center(M, '-'))
    
#     # Middle part
#     print('WELCOME'.center(M, '-'))
    
#     # Bottom part (same as top part but reversed)
#     for i in range(N // 2 - 1, -1, -1):
#         pattern = '.|.' * (2 * i + 1)
#         print(pattern.center(M, '-'))

# if __name__ == '__main__':
#     N, M = map(int, input().split())
#     print_door_mat(N, M)



# # Given an integer,n , print the following values for each integer i from 1 to n:

# # Decimal
# # Octal
# # Hexadecimal (capitalized)
# # Binary
# # Function Description

# # Complete the print_formatted function in the editor below.

# # print_formatted has the following parameters:

# # int number: the maximum value to print
# # Prints

# # The four values must be printed on a single line in the order specified above for each i from 1 to number.
# #  Each value should be space-padded to match the width of the binary value of  number and
# # the values should be separated by a single space.


# def print_formatted(number):
#     width = len(bin(number)[2:])
    
#     # Iterate from 1 to number
#     for i in range(1, number + 1):
#         decimal_str = format(i, 'd').rjust(width)  # Decimal
#         octal_str = format(i, 'o').rjust(width)    # Octal
#         hex_str = format(i, 'X').rjust(width)      # Capitalized Hexadecimal
#         binary_str = format(i, 'b').rjust(width)   # Binary
        
#         # Print all formatted values in a single line
#         print(f"{decimal_str} {octal_str} {hex_str} {binary_str}")

# if __name__ == '__main__':
#     n = int(input())
#     print_formatted(n)




# # You are given an integer, . Your task is to print an alphabet rangoli of size . 
# # (Rangoli is a form of Indian folk art based on creation of patterns.)
# # Different sizes of alphabet rangoli are shown below:

# # #size 3
# # ----c----
# # --c-b-c--
# # c-b-a-b-c
# # --c-b-c--
# # ----c----

# # #size 5

# # --------e--------
# # ------e-d-e------
# # ----e-d-c-d-e----
# # --e-d-c-b-c-d-e--
# # e-d-c-b-a-b-c-d-e
# # --e-d-c-b-c-d-e--
# # ----e-d-c-d-e----
# # ------e-d-e------
# # --------e--------

# # #size 10

# # ------------------j------------------
# # ----------------j-i-j----------------
# # --------------j-i-h-i-j--------------
# # ------------j-i-h-g-h-i-j------------
# # ----------j-i-h-g-f-g-h-i-j----------
# # --------j-i-h-g-f-e-f-g-h-i-j--------
# # ------j-i-h-g-f-e-d-e-f-g-h-i-j------
# # ----j-i-h-g-f-e-d-c-d-e-f-g-h-i-j----
# # --j-i-h-g-f-e-d-c-b-c-d-e-f-g-h-i-j--
# # j-i-h-g-f-e-d-c-b-a-b-c-d-e-f-g-h-i-j
# # --j-i-h-g-f-e-d-c-b-c-d-e-f-g-h-i-j--
# # ----j-i-h-g-f-e-d-c-d-e-f-g-h-i-j----
# # ------j-i-h-g-f-e-d-e-f-g-h-i-j------
# # --------j-i-h-g-f-e-f-g-h-i-j--------
# # ----------j-i-h-g-f-g-h-i-j----------
# # ------------j-i-h-g-h-i-j------------
# # --------------j-i-h-i-j--------------
# # ----------------j-i-j----------------
# # ------------------j------------------



# def print_rangoli(size):
#     import string
#     alphabet = string.ascii_lowercase
    
#     # Determine the maximum character
#     max_char = alphabet[size - 1]
    
#     # Build the rangoli
#     lines = []
#     for i in range(size):
#         # Create the sequence of characters
#         line = '-'.join(alphabet[size - 1:i:-1] + alphabet[i:size])
#         lines.append(line.center(4 * size - 3, '-'))
    
#     # Print the rangoli
#     for line in lines[::-1] + lines[1:]:
#         print(line)

# if __name__ == '__main__':
#     n = int(input())
#     print_rangoli(n)





# # Kevin and Stuart want to play the 'The Minion Game'.

# # Game Rules

# # Both players are given the same string, .
# # Both players have to make substrings using the letters of the string .
# # Stuart has to make words starting with consonants.
# # Kevin has to make words starting with vowels.
# # The game ends when both players have made all possible substrings.

# # Scoring
# # A player gets +1 point for each occurrence of the substring in the string .

# # For Example:
# # String  = BANANA
# # Kevin's vowel beginning word = ANA
# # Here, ANA occurs twice in BANANA. Hence, Kevin will get 2 Points.
# # For better understanding, see the image below:
# # banana.png
# # Your task is to determine the winner of the game and their score.

# # Function Description
# # Complete the minion_game in the editor below.
# # minion_game has the following parameters:
# # string string: the string to analyze
# # Prints

# # string: the winner's name and score, separated by a space on one line, or Draw if there is no winner


# def minion_game(string):
#     vowels = 'AEIOU'
#     length = len(string)
#     stuart_score = 0
#     kevin_score = 0
    
#     for i in range(length):
#         if string[i] in vowels:
#             kevin_score += (length - i)
#         else:
#             stuart_score += (length - i)
    
#     if stuart_score > kevin_score:
#         print(f"Stuart {stuart_score}")
#     elif kevin_score > stuart_score:
#         print(f"Kevin {kevin_score}")
#     else:
#         print("Draw")

# if __name__ == '__main__':
#     s = input()
#     minion_game(s)








# # You are asked to ensure that the first and last names of people begin with a capital letter in their passports. For example, alison heck should be capitalised correctly as Alison Heck.
# # alison heck => Alison Heck
# # Given a full name, your task is to capitalize the name appropriately.

# # Input Format
# # A single line of input containing the full name, S.

# # Constraints
# # 0<len(S) < 1000

# # The string consists of alphanumeric characters and spaces.
# # Note: in a word only the first character is capitalized. Example 12abc when capitalized remains 12abc.

# # Output Format
# # Print the capitalized string, S.

# # Sample Input
# # chris alan

# # Sample Output
# # Chris Alan



# #!/bin/python3

# import math
# import os
# import random
# import re
# import sys

# # Complete the solve function below.
# def solve(s):
#     # Capitalize each word in the string
#     capitalized_string = ' '.join(word.capitalize() for word in s.split(' '))
#     return capitalized_string


# if __name__ == '__main__':
#     fptr = open(os.environ['OUTPUT_PATH'], 'w')

#     s = input()

#     result = solve(s)

#     fptr.write(result + '\n')

#     fptr.close()








# # In Python, a string of text can be aligned left, right and center.
# # .ljust(width)

# # This method returns a left aligned string of length width.
# # >>> width = 20
# # >>> print 'HackerRank'.ljust(width,'-')
# # HackerRank----------  
# # .center(width)

# # This method returns a centered string of length width.
# # >>> width = 20
# # >>> print 'HackerRank'.center(width,'-')
# # -----HackerRank-----
# # .rjust(width)

# # This method returns a right aligned string of length width.
# # >>> width = 20
# # >>> print 'HackerRank'.rjust(width,'-')
# # ----------HackerRank

# # Task
# # You are given a partial code that is used for generating the HackerRank Logo of variable thickness.
# # Your task is to replace the blank (______) with rjust, ljust or center.

# # Input Format
# # A single line containing the thickness value for the logo.

# # Constraints
# # The thickness must be an odd number.
# # 0<thickness<50

# # Output Format
# # Output the desired logo.

# # Sample Input
# # 5

# # Sample Output
# #     H    
# #    HHH   
# #   HHHHH  
# #  HHHHHHH 
# # HHHHHHHHH
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #   HHHHHHHHHHHHHHHHHHHHHHHHH   
# #   HHHHHHHHHHHHHHHHHHHHHHHHH   
# #   HHHHHHHHHHHHHHHHHHHHHHHHH   
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #   HHHHH               HHHHH             
# #                     HHHHHHHHH 
# #                      HHHHHHH  
# #                       HHHHH   
# #                        HHH    
# #                         H 
# # Language
# # Pypy 3
# # More
# # 1
# # # Enter your code here. Read input from STDIN. Print output to STDOUT
# # Line: 1 Col: 70

# # Test against custom input




# # Enter your code here. Read input from STDIN. Print output to STDOUT
# thickness = int(input()) # This must be an odd number
# c = 'H'

# # Top Cone
# for i in range(thickness):
#     print((c*i).rjust(thickness-1) + c + (c*i).ljust(thickness-1))

# # Top Pillars
# for i in range(thickness+1):
#     print((c*thickness).center(thickness*2) + (c*thickness).center(thickness*6))

# # Middle Belt
# for i in range((thickness+1)//2):
#     print((c*thickness*5).center(thickness*6))    

# # Bottom Pillars
# for i in range(thickness+1):
#     print((c*thickness).center(thickness*2) + (c*thickness).center(thickness*6))    

# # Bottom Cone
# for i in range(thickness):
#     print(((c*(thickness-i-1)).rjust(thickness) + c + (c*(thickness-i-1)).ljust(thickness)).rjust(thickness*6))
